### PR TITLE
Change training stage from ResultStage to ShuffleMapStage

### PR DESCRIPTION
**Pull request by jinmfeng001**
The training step runs in barrier model and is ResultStage, which means when one of the training task failed the training stage failed, and the whole application failed becuase no retry for ResultStage. 
When there are many training tasks, it's easy that one task will fail due to cluster node issue, so the application may has a high failure rate. 
In this PR, we add a repartition step to make the training stage as ShuffleMapStage, so that when the training stage failed, the stage can retry and the whole spark application won't fail.

### CODE REVIEW

**Comment by wbo4958** [comment]
To fix https://github.com/dmlc/xgboost/issues/8336

**Comment by wbo4958** [comment]
@jinmfeng001, Per my understanding, the xgboost training is deterministic, @trivialfis please correct me. if the first training fails, the retry, I guess, may still fail (the only difference I can think of is the training task may be distributed to different workers). 

BTW, have you seen the retry can really help the xgboost training?

**Comment by trivialfis** [comment]
Indeed, there can be many reasons for failure and xgboost doesn't do well on large shared clusters due to the lack of a failure recovery mechanism.

@wbo4958 Please help review the PR. In addition, is there a way to test?

**Comment by jinmfeng001** [comment]
> Indeed, there can be many reasons for failure and xgboost doesn't do well on large shared clusters due to the lack of a failure recovery mechanism.
> 
> @wbo4958 Please help review the PR. In addition, is there a way to test?

In our situation, we changed it to ShuffleMapStage, and we don't see much application failures now. 
It's a good question how to test this change. I don't there's any thing we can do to fail a task. Do you guys know?

**Comment by wbo4958** [comment]
@jinmfeng001 would you like to provide the whole log with the retry working?

**Comment by jinmfeng001** [comment]
> 
We're focusing on resolving the failure cuased by hardware failure on one node(In a big cluster, it's common one node may fail due to some hardware issue). So if the task is reassigned to another node it may succeed. 

In our situation, this help to remove the failure rate for the whol training applciation.

**Comment by wbo4958** [comment]
@jinmfeng001 Thx for your explanation, just give me some time to learn it.

**Comment by wbo4958** [comment]
@jinmfeng001 Thx for your PR, I spent some time learning the spark's retry mechanism. You're right, the spark won't retry the failed barrier ResultStage. 

I don't know if this PR will introduce the performance issue since another shuffle stage will involve the disk write/read. I will do some test. Thx for your time.

**Comment by wbo4958** [comment]
I ran the mortgage test with xgboost-spark-gpu on the spark standalone cluster with 1 worker and spark-rapids enabled. Here is the result


||1st|2nd|3rd|
|---|---|---|---|
|W/O PR|89.379|89.126| 89.367|
|W/ PR|90.173 | 89.224 | 89.131 |

Seems this PR doesn't introduce much extra overhead.

**Review by wbo4958 — APPROVED** [review]

**Comment by wbo4958** [comment]
LGTM

**Review by trivialfis — APPROVED** [review]
